sql: prevent dropping a sequence if it's being used

vilterp opened the pull request:
Save and update references between sequence descriptor and column descriptor; check these references when trying to drop a sequence.

Addresses #20964

Commit:
sql: store sequence <=> column references on CREATE TABLE

Release note: None

Commit:
sql: prevent dropping a sequence if a table uses it

Release note: None

Commit:
sql: update sequence references on DROP COLUMN

If a column which was using a sequence is dropped, remove the reference
from the sequence descriptor to the dropped column.

Release note: None

Commit:
sql: update sequence references on ADD COLUMN

Release note: None

Commit:
sql: update sequence references on SET DEFAULT / DROP DEFAULT

Changing the default expression of a column can add and/or remove a
reference between a column descriptor and a sequence descriptor. This PR
ensures that they stay up to date.

Release note: None

Commit:
sql: add sequences to crdb_internal.{backward, forward}_dependencies

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21364)
<!-- Reviewable:end -->

vilterp commented:
Requesting review from @jordanlewis and @vivekmenezes because it does a lot of modification of table descriptors, and I may have made mistakes with how those changes are saved / notified.

jordanlewis commented:
Taking a look at this today. Sorry for the delay!

jordanlewis commented:
Generally looks good! The main question on my mind is about defaults that have more than one `nextval` call which doesn't seem to be properly handled.

---

Reviewed 5 of 8 files at r1, 1 of 2 files at r2, 1 of 3 files at r3, 2 of 2 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sequence.go, line 241 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eSuJJmi72oKso9Fb:-L35eSuKONPPejBIO0Pw:b-n8h7tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L241)):*
> ```Go
> 	}
> 	col.UsesSequenceId = &seqDesc.ID
> 	// Now that we've gotten the sequence descriptor, I guess we can mutate it?
> ```

Probably should change this comment to be less questiony :)

---

*[pkg/sql/sequence.go, line 289 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eYY7_m7V_P0W2Z28:-L35eYY7_m7V_P0W2Z29:bo3j0vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L289)):*
> ```Go
> // no call to nextval.
> // e.g. nextval('foo') => "foo"; <some other expression> => nil
> func getUsedSequenceName(defaultExpr tree.TypedExpr) (*string, error) {
> ```

What if there is more than one nextval in the expression?

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 597 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35enpgbP3uSTaLyijL:-L35enpgbP3uSTaLyijM:b-ygdpt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/logictest/testdata/logic_test/sequences#L597)):*
> ```
> 
> statement ok
> ALTER TABLE drop_default_tesT_tbl ALTER COLUMN id SET DEFAULT nextval('drop_default_test_seq_2')
> ```

What about some tests that we can drop the sequence after the entire table is dropped, not just the column?

---

*[pkg/sql/sqlbase/structured.proto, line 131 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35ebUug8PkIy4eZN-9:-L35ebUug8PkIy4eZN-A:bpbv2ko) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/structured.proto#L131)):*
> ```protobuf
>   // Id of a sequence used in this column's DEFAULT expression, in a call to nextval().
>   // null if no sequence is used.
>   optional uint32 uses_sequence_id = 10 [(gogoproto.nullable) = true, (gogoproto.casttype) = "ID"];
> ```

Probably needs to be a repeated field.

---

*[pkg/sql/sqlbase/table.go, line 157 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eeZCangp2A97vsUi:-L35eeZCangp2A97vsUj:b-xokrx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/table.go#L157)):*
> ```Go
> func MakeColumnDefDescs(
> 	d *tree.ColumnTableDef, semaCtx *tree.SemaContext, evalCtx *tree.EvalContext,
> ) (*ColumnDescriptor, *IndexDescriptor, tree.TypedExpr, error) {
> ```

Please add a comment to explain the new return parameter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-:-L35f-Zr8wrjj43AlDTd:bn6vji1)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sequence.go:
```diff
@@ -208,3 +209,109 @@ func assignSequenceOptions(
+	col *sqlbase.ColumnDescriptor,
+	expr tree.TypedExpr,
+	evalCtx *tree.EvalContext,
+) (*sqlbase.TableDescriptor, error) {
```
Can you add a comment on what descriptor is returned?

vivekmenezes commented on pkg/sql/sequence.go:
```diff
@@ -208,3 +209,109 @@ func assignSequenceOptions(
+	// Now that we've gotten the sequence descriptor, I guess we can mutate it?
+	seqDesc.DependedOnBy = append(seqDesc.DependedOnBy, sqlbase.TableDescriptor_Reference{
+		ID:        tableDesc.ID,
+		ColumnIDs: []sqlbase.ColumnID{col.ID},
```
what if there are multiple columns using the same sequence?

vilterp replied:
`maybeAddSequenceDependency` gets called for each column, so it'll end up as a separate entry in `seqDesc.DependedOnBy` for each. Will add a logic test for this.

vivekmenezes commented on pkg/sql/sequence.go:
```diff
@@ -208,3 +209,109 @@ func assignSequenceOptions(
+	}
+	// Find an item in seqDesc.DependedOnBy which references tableDesc.
+	refIdx := -1
+	for i := 0; i < len(seqDesc.DependedOnBy); i++ {
```
for i, reference := range ...

vilterp replied:
Done

vivekmenezes commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -126,6 +126,9 @@ message ColumnDescriptor {
   reserved 7;
+  // Id of a sequence used in this column's DEFAULT expression, in a call to nextval().
+  // null if no sequence is used.
+  optional uint32 uses_sequence_id = 10 [(gogoproto.nullable) = true, (gogoproto.casttype) = "ID"];
```
Is it possible for more than one sequence to be used?

vivekmenezes reviewed (commented):
LGTM

Commit:
sql: record deps for multiple calls to nextval() in DEFAULT expression

In case someone does e.g. `DEFAULT nextval('seq_a') + nextval('seq_b')`.

Release note: None

Commit:
sql: update sequence references on DROP TABLE

Release note: None

vilterp commented:
TFTRs. PTAL

---

Review status: 1 of 15 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sequence.go, line 241 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eSuJJmi72oKso9Fb:-L3A4pHEgCbk6lVVYJ9j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L241)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Probably should change this comment to be less questiony :)
</blockquote></details>

Done.

---

*[pkg/sql/sequence.go, line 289 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eYY7_m7V_P0W2Z28:-L3A4q_r0xisa6R5OSi6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sequence.go#L289)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What if there is more than one nextval in the expression?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 597 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35enpgbP3uSTaLyijL:-L3A4rZ3vdKqhXgjZtvS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/logictest/testdata/logic_test/sequences#L597)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What about some tests that we can drop the sequence after the entire table is dropped, not just the column?
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 131 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35ebUug8PkIy4eZN-9:-L3A4sOrRIulRsgeOK0K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/structured.proto#L131)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Probably needs to be a repeated field.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 157 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-L35eeZCangp2A97vsUi:-L3A4tYQ_CzoD_DqFmtw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f409aba851714c8f7be89705b81f727819749c1/pkg/sql/sqlbase/table.go#L157)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Please add a comment to explain the new return parameter.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21364#-:-L3A4vqeWEZwuyQyuhU9:bqslf27)*
<!-- Sent from Reviewable.io -->

knz commented:
nit: it's frankly incorrect to not have entered any release note in any of these commits. Here a pretty significant bug was fixed and it would have deserved mention in the next batch of release notes.

Adding the label `docs-todo` as a reminder to have this PR properly documented in the next release notes despite the lack of release note annotation. cc @jseldess .

jseldess commented:
Thanks, @knz. Tracked here: https://github.com/cockroachdb/docs/issues/2409
